types/canonical.go

```diff
@@ -51,26 +51,30 @@ func CanonicalizeProposal(chainID string, proposal *tmproto.Proposal) tmproto.Ca
 	}
 }
 
-func GetVoteExtensionToSign(ext *tmproto.VoteExtension) *tmproto.VoteExtensionToSign {
-	if ext == nil {
-		return nil
-	}
-	return &tmproto.VoteExtensionToSign{
-		AppDataToSign: ext.AppDataToSign,
-	}
-}
-
 // CanonicalizeVote transforms the given Vote to a CanonicalVote, which does
-// not contain ValidatorIndex and ValidatorAddress fields.
+// not contain ValidatorIndex and ValidatorAddress fields, or any fields
+// relating to vote extensions.
 func CanonicalizeVote(chainID string, vote *tmproto.Vote) tmproto.CanonicalVote {
 	return tmproto.CanonicalVote{
-		Type:          vote.Type,
-		Height:        vote.Height,       // encoded as sfixed64
-		Round:         int64(vote.Round), // encoded as sfixed64
-		BlockID:       CanonicalizeBlockID(vote.BlockID),
-		Timestamp:     vote.Timestamp,
-		ChainID:       chainID,
-		VoteExtension: GetVoteExtensionToSign(vote.VoteExtension),
+		Type:      vote.Type,
+		Height:    vote.Height,       // encoded as sfixed64
+		Round:     int64(vote.Round), // encoded as sfixed64
+		BlockID:   CanonicalizeBlockID(vote.BlockID),
+		Timestamp: vote.Timestamp,
+		ChainID:   chainID,
+	}
+}
+
+// CanonicalizeVoteExtension extracts the vote extension from the given vote
+// and constructs a CanonicalizeVoteExtension struct, whose representation in
+// bytes is what is signed in order to produce the vote extension's signature.
+func CanonicalizeVoteExtension(chainID string, vote *tmproto.Vote) tmproto.CanonicalVoteExtension {
+	return tmproto.CanonicalVoteExtension{
+		Extension: vote.Extension,
+		Height:    vote.Height,
+		Round:     vote.Round,
```

Comment:
Silly question :-)
In line 61, the round is cast to `int64`, whereas we don't do it here. Should we also cast to `int64` here? Or is the cast in line 61 actually unnecessary?

Reply:
`Round` here is modeled as `int32`. Do we anticipate the round number to surpass 2 billion?

Reply:
Not 100% sure whether int32 or int64 (2 billion is a lot, I agree).
However, shouldn't we be consistent between `CanonicalVote` and `CanonicalVoteExtension` regarding the type of `Round`?.
Probably int32 is more reasonable, but since `CanonicalVote` has been around for a long time, if we change it now to int32, we might risk breaking signature verification for past blocks in existing chains.
Thoughts?

Reply:
Ah, I see that it's actually modeled like this because we require fixed-size integers to ensure canonicalization: https://github.com/tendermint/tendermint/blob/68c9efdb825e95a2d2c1f4e6cb5dc93082b3a6dd/proto/tendermint/types/canonical.proto#L33

I'll change the code vote `CanonicalVoteExtension` accordingly.

Reply:
Thinking about this again now, I don't really understand why we need fixed-size integers, or why `sfixed32` wasn't used for the round number, for canonicalization. Unless the serialization is somehow nondeterministic with non-fixed-sized integer values?

Reply:
They're just signed fixed-length 32- and 64-bit integers, respectively: https://developers.google.com/protocol-buffers/docs/proto3#scalar

The other non-fixed-length integers are encoded using variable-length encoding.